db/core/src/main/java/org/ethereum/beacon/db/source/HoleyList.java

```diff
@@ -0,0 +1,59 @@
+package org.ethereum.beacon.db.source;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Add-only list which can miss elements at some positions and its size is the maximal element index
+ * Also can be treated as <code>Map&lt;Long, V&gt;</code> with maximal key tracking.
+ */
+public interface HoleyList<V> {
+
+  /**
+   * Maximal index of inserted element + 1
+   */
+  long size();
+
+  /**
+   * Put element at index <code>idx</code>
+   * Increases size if necessary
+   * If value is null nothing is modified
+   */
+  void put(long idx, V value);
+
+  /**
+   * Returns element at index <code>idx</code>
+   * Empty instance is returned if no element with this index
+   */
+  Optional<V> get(long idx);
+
+  /**
+   * Puts element with index <code>size()</code>
+   */
+  default void add(V value) {
+    put(size(), value);
```

Comment:
I'd return result here

Reply:
What result?

Reply:
Could it update old one, or it's always new?

Reply:
Yep, it's always new